src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+	uint8_t brightness;
+	struct gskill_led_color dpi[4];
+	struct gskill_led_color leds[GSKILL_LED_TYPE_COUNT];
+} __attribute__((packed));
+
+struct gskill_dpi_led_group_cfg {
+	uint8_t duration_step;
+	uint8_t duration_high;
+	uint8_t duration_low;
+	uint8_t cycle_num;
+
+	struct gskill_led_values steps[12];
+};
+
+struct gskill_led_group_cfg {
+	enum gskill_led_control_type type :3;
+	uint8_t                           :5; /* unused */
+
+	uint8_t duration_step;
+	uint8_t duration_high;
+	uint8_t duration_low;
+	uint8_t cycle_num;
+
+	struct gskill_led_values steps[12];
+} __attribute__((packed));
+
+struct gskill_dpi_led_cycle_cfg {
+	enum gskill_led_control_type type :3;
+	uint8_t                           :5; /* unused */
+
+	/* Don't worry, the low/high flip-flop here is intentional */
+	uint8_t duration_low;
+	uint8_t duration_high;
+	uint8_t cycle_num;
+
+	struct gskill_led_values cycles[12];
+} __attribute__((packed));
+
+/*
+ * We may occasionally run into codes outside this, however those codes
+ * indicate functionalities that aren't too useful for us
+ */
+enum gskill_button_function_type {
+	GSKILL_BUTTON_FUNCTION_WHEEL                = 0x00,
+	GSKILL_BUTTON_FUNCTION_MOUSE                = 0x01,
+	GSKILL_BUTTON_FUNCTION_KBD                  = 0x02,
+	GSKILL_BUTTON_FUNCTION_CONSUMER             = 0x03,
+	GSKILL_BUTTON_FUNCTION_MACRO                = 0x06,
+	GSKILL_BUTTON_FUNCTION_DPI_UP               = 0x09,
+	GSKILL_BUTTON_FUNCTION_DPI_DOWN             = 0x0a,
+	GSKILL_BUTTON_FUNCTION_CYCLE_DPI_UP         = 0x0b,
+	GSKILL_BUTTON_FUNCTION_CYCLE_DPI_DOWN       = 0x0c,
+	GSKILL_BUTTON_FUNCTION_PROFILE_SWITCH       = 0x0d,
+	GSKILL_BUTTON_FUNCTION_TEMPORARY_CPI_ADJUST = 0x15,
+	GSKILL_BUTTON_FUNCTION_DIRECT_DPI_CHANGE    = 0x16,
+	GSKILL_BUTTON_FUNCTION_CYCLE_PROFILE_UP     = 0x18,
+	GSKILL_BUTTON_FUNCTION_CYCLE_PROFILE_DOWN   = 0x19,
+	GSKILL_BUTTON_FUNCTION_DISABLE              = 0xff
+};
+
+struct gskill_button_cfg {
+	enum gskill_button_function_type type :8;
+	union {
+		struct {
+			enum {
+				GSKILL_WHEEL_SCROLL_UP = 0,
+				GSKILL_WHEEL_SCROLL_DOWN = 1,
+			} direction :8;
+		} wheel;
+
+		struct {
+			enum {
+				GSKILL_BTN_MASK_LEFT   = (1 << 0),
+				GSKILL_BTN_MASK_RIGHT  = (1 << 1),
+				GSKILL_BTN_MASK_MIDDLE = (1 << 2),
+				GSKILL_BTN_MASK_SIDE   = (1 << 3),
+				GSKILL_BTN_MASK_EXTRA  = (1 << 4)
+			} button_mask :8;
+		} mouse;
+
+		struct {
+			uint16_t code;
+		} consumer;
+
+		struct {
+			uint8_t modifier_mask;
+			uint8_t hid_code;
+			/*
+			 * XXX: Supposedly this is supposed to have additional
+			 * parts of the kbd code, however that doesn't seem to
+			 * be the case in practice…
+			 */
+			uint16_t :16;
+		} kbd;
+
+		struct {
+			uint8_t level;
+		} dpi;
+	} params;
+} __attribute__((packed));
+
+struct gskill_action_mapping {
+	struct gskill_button_cfg config;
+	struct ratbag_button_action action;
+};
+
+struct gskill_profile_report {
+	uint16_t                  :16;
+	uint8_t profile_num;
```

Comment:
From what I quickly read on the spec, the profile_num is in the second byte, not the third (when doing profile read).